src/server/gc_worker/gc_worker.rs

```diff
@@ -727,15 +800,31 @@ where
         max_ts: TimeStamp,
         start_key: &Key,
         limit: usize,
+        regions_provider: Arc<dyn RegionInfoProvider>,
     ) -> Result<Vec<LockInfo>> {
-        let snap = self
-            .engine
-            .snapshot_on_kv_engine(start_key.as_encoded(), &[])
-            .unwrap();
-        let mut reader = MvccReader::new(snap, Some(ScanMode::Forward), false);
-        let (locks, _) = reader
-            .scan_locks(Some(start_key), None, |l| l.ts <= max_ts, limit)
-            .map_err(TxnError::from_mvcc)?;
+        let regions = box_try!(regions_provider.get_regions_in_range(start_key.as_encoded(), &[]))
+            .into_iter()
+            .filter(move |r| find_peer(r, self.store_id).is_some());
+
+        let mut first_round = true;
+        let mut locks = Vec::new();
+        for region in regions {
+            let start_key = {
+                if first_round {
+                    first_round = false;
+                    start_key.clone()
+                } else {
+                    Key::from_raw(region.get_start_key())
+                }
+            };
+            let snap = self.get_snapshot(self.store_id, &region)?;
+            let mut reader = MvccReader::new(snap, Some(ScanMode::Forward), false);
+            let (locks_this_region, _) = reader
+                .scan_locks(Some(&start_key), None, |l| l.ts <= max_ts, limit)
```

Comment:
/cc @lhy1024 Will it impact the statistics? As there will be many scan ops and end key is always None.

Reply:
Scan is also limited by region boundries of snapshot.

Reply:
If this pr only affects gc, it has no effect on hotspot statistics, because pd does not process gc-related statistics for the time being